### ISSUE TITLE
Matrix Nets: A New Deep Architecture for Object Detection

### ISSUE THREAD

**Issue by WongKinYiu** [AlexeyAB/darknet#3772]
Could this repo supports max pooling layer with different x, y strides.
I would like to implement the state-of-the-art object detector.
Thanks.

![image](https://user-images.githubusercontent.com/12152972/63207472-b22a0a00-c0f9-11e9-9004-51bbeeb71b9d.png)

**Comment by AlexeyAB** [comment]
https://arxiv.org/abs/1908.04646v2

![image](https://user-images.githubusercontent.com/4096485/63209321-e8f72480-c0e7-11e9-8ab5-b75702dfcd29.png)



 xNets map objects with different sizes and aspect ratios into layers where the sizes and the aspect ratios of the objects within their layers are nearly uniform. Hence, xNets provide a scale and aspect ratio aware architecture. We leverage xNets to enhance key-points based object detection. Our architecture achieves **mAP of 47.8 on MS COCO**, which is higher than any other single-shot detector while using half the number of parameters and training 3x faster than the next best architecture.

**Comment by AlexeyAB** [comment]
> Could this repo supports max pooling layer with different x, y strides.

Is this the only necessary feature for the implementation of the Matrix is not?
So we should have:
```
stride_x=
stride_y=
```
Or do you need something else?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,
Yes, but if it can support convolutional layer & average pooling layer with stride_x and stride_y is better.
The paper seems use convolutional layers to do down-sampling.

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi,

I added support for
```
[maxpool]
stride_x=2
stride_y=3
...
```

**Comment by WongKinYiu** [comment]
@AlexeyAB  Hello, thank you very very much.

**Comment by AlexeyAB** [comment]
It seems that MatrixNet (different strides) and TridentNet (different dilations) are very promising approaches for generalizing different sizes and aspect ratios of objects.

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi,

What progress?

I added stride_x=... stride_y=... for convolutional layer. So you can try to make MatrixNet.

```
[convolutional]
stride_x=2
stride_y=3
...
```

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello, for the version with different stride max pooling layers, now training 200k epochs.
it needs 7~9 more days to finish training.

**Comment by LukeAI** [comment]
@WongKinYiu  are you training a matrixnet with the original resnext-101 backbone? or something else?

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks, do you know about "shared 3x3 convolutions"? https://github.com/AlexeyAB/darknet/issues/3772#issuecomment-526918433

**Comment by WongKinYiu** [comment]
@AlexeyAB Yes, I think it similar to TridentNet.
In my understand, only layers with the same aspect-ratio share the same weights.
But for max pooling version, there is no parameters.

**Comment by WongKinYiu** [comment]
@LukeAI Hello.
I do not use big model like resnet-50, resnext-101...
I only train the very small models.

**Comment by LukeAI** [comment]
which one? :)

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks. 

You can try to train 3 models:
1. with `[maxpool] stride_x=... stride_y=...` 

2. with `[convolutional] stride_x=... stride_y=...`  with `share_index=...` in some layers, to get weights from conv-layer with this number (as in TridentNet)

3. then with `[convolutional] stride_x=... stride_y=... antialiasing=1` with `share_index=...` in some layers


And compare results.

I added `antialiasing=1` parameter for `[convolutional]` layer: https://github.com/AlexeyAB/darknet/issues/3672

**Comment by WongKinYiu** [comment]
@AlexeyAB Thank you for your advise. I will get free gpus after two weeks.

**Comment by WongKinYiu** [comment]
@AlexeyAB 

maxpool version:

| Model | +anchors | BFLOPs | mAP@0.5 | mAP@0.5:0.95 |
| :-- | :-: | :--: | :-: | :-: |
| original | - | 5.32 | 48.7 | 25.7 |
| matrix net | no | 5.22 | 49.1 | 25.4 |
| matrix net | yes | - | - | - |

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks for results!
* What does it mean anchors - no? Did you implement [yolo]-layers without anchors?
* Can you share cfg-file?

**Comment by WongKinYiu** [comment]
@AlexeyAB 

It means without increase number of anchors.
![image](https://user-images.githubusercontent.com/12152972/64473494-396b1a80-d19a-11e9-9730-667eb7733b51.png)

**Comment by AlexeyAB** [comment]
@WongKinYiu Is Original - original MatrixNet from https://arxiv.org/abs/1908.04646v2 ?

**Comment by WongKinYiu** [comment]
@AlexeyAB 

No, original means a yolov3-based model, without adding the feature proposed by MatrixNet.

**Comment by WongKinYiu** [comment]
cfg files based on yolov3-tiny

[yolov3-tiny_3l(15.778BFLOPs).cfg.txt](https://github.com/AlexeyAB/darknet/files/3587556/yolov3-tiny_3l.15.778BFLOPs.cfg.txt)
[yolov3-tiny_3l_maxpoolmatrixnet(15.565BFLOPs).cfg.txt](https://github.com/AlexeyAB/darknet/files/3587557/yolov3-tiny_3l_maxpoolmatrixnet.15.565BFLOPs.cfg.txt)
[yolov3-tiny_3l_maxpoolmatrixnet_addanchors(15.787BFLOPs).cfg.txt](https://github.com/AlexeyAB/darknet/files/3587558/yolov3-tiny_3l_maxpoolmatrixnet_addanchors.15.787BFLOPs.cfg.txt)

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks. What mAP@0.5 did you get for yolov3-tiny_3l_maxpoolmatrixnet_addanchors(15.787BFLOPs).cfg.txt ?

**Comment by AlexeyAB** [comment]
@WongKinYiu 

Did you anywhere meet original MatrixNet-backbone (not just yolov3/tiny with non-unoform strides), if yes - can you share it?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

I do not meet original MatrixNet-backbone.

In the paper, they only show the concept of MatrixNet and use it on CornerNet without telling details.
But the concept can be used on YOLO.
![image](https://user-images.githubusercontent.com/12152972/64495527-f0b47e00-d2cd-11e9-98db-36b89c0d2026.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB i v updated the results of https://github.com/AlexeyAB/darknet/issues/3772#issuecomment-529091355

**Comment by AlexeyAB** [comment]
@WongKinYiu Do you plane to try to use non-uniform strided x/y conv layers with shared weights?

**Comment by WongKinYiu** [comment]
@AlexeyAB Yes, i will get available gpus after 1~2 weeks.

And I would like to solve the problem https://github.com/AlexeyAB/darknet/issues/3708#issuecomment-528140698 first.
But there are too much commits to be check.

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi, did you try non-uniform strided x/y conv layers with shared weights?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

yes, it is slightly better on mAP@0.5, but decreases mAP@0.5:0.95.
i modified the architectures to make it has better results.
but for my project, the FPS drops too much.
![image](https://user-images.githubusercontent.com/12152972/67152434-d0360580-f308-11e9-8824-c801d707c517.png)

**Comment by AlexeyAB** [comment]
@WongKinYiu 

* Is MatrixNet - non-unoform conv-layers?

* Is Modified MatrixNet - non-unoform maxpool-layers?

**Comment by WongKinYiu** [comment]
![image](https://user-images.githubusercontent.com/12152972/67153048-869fe780-f315-11e9-83ff-ce7a74708946.png)
i apply concatenate (route layer) for the blocks with multiple in-degree.

**Comment by WongKinYiu** [comment]
And, for YOLO, we can simply add number of anchors to achieve high mAP@0.5:0.95.
![image](https://user-images.githubusercontent.com/12152972/67153313-68d48180-f319-11e9-96f1-db01e4cc4917.png)

**Comment by AlexeyAB** [comment]
@WongKinYiu Can you share the last cfg/weights-file?

**Comment by WongKinYiu** [comment]
sorry for that i can not share the cfg file of our backbone due to non-disclosure agreement.
here are the cfg file of the heads, i do not use share weights in my implementation.

1. implemented matrixnet - [matrixnet-9anchors.txt](https://github.com/AlexeyAB/darknet/files/3747764/matrixnet-9anchors.txt)

2. implemented modified matrixnet - [mmatrixnet-9anchors.txt](https://github.com/AlexeyAB/darknet/files/3747765/mmatrixnet-9anchors.txt)

3. implemented modified matrixnet with more anchors - [mmatrixnet-25anchors.txt](https://github.com/AlexeyAB/darknet/files/3747766/mmatrixnet-25anchors.txt)

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks! 

* Do you use light custom backbone - something like tinyv3/prn/...?
* Did this mAP get on MS COCO?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

* I use Pelee-based backbone, the FPS is testing on GTX 1080ti, so I think it is about 40% faster than FPS on GTX titan X.
* If using our new **darknet reference + prn**-based model, I thank the mAP will drop 3~5%, but the FPS will be twice.
* All mAP get on MS COCO test-dev.

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi,
I just temporary changed [yolo] layer for training.
Now it uses `ignore_thresh` only if class_id's matched: https://github.com/AlexeyAB/darknet/commit/e6486ab594e877e0b870eab6788de9e888c35840#diff-180a7a56172e12a8b79e41ec95ae569dR316

So you can try to train one of model with these changes.

**Comment by WongKinYiu** [comment]
@AlexeyAB thanks,
i will get a free gpu at Thursday.

**Comment by VideoPlayerCode** [comment]
Very interesting. Any idea why mAP did not improve when using matrix architecture?

**Comment by AlexeyAB** [comment]
@WongKinYiu @VideoPlayerCode 
May be there are used not the most suitable anchors for different [yolo] layers.

May be we should use synthetic anchors like

`anchors = 10,10,  30,15,  15,30,  30,30,  60,30,  30,60,  60,60, 120,60,  60,120, 120,120,  240,120,  120,240,  240,240`
`num=13`

instead of
`anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326`
`num=9`

**Comment by AlexeyAB** [comment]
@WongKinYiu I added model ` yolo_v3_tiny_pan3 matrix_gaussian aa_ae_mixup.cfg` 

* based on yolov3-tiny, so can be trained by using `yolov3-tiny.conv.15` pre-trained weights
* + pan3 block
* + modified your matrix head `mmatrixnet-25anchors.txt` https://github.com/AlexeyAB/darknet/issues/3772#issuecomment-544242134
* by using `[Gaussian_yolo]` layer instead of `[yolo]`
* with AntiAliasing, Assisted Excitation and Mixup

https://github.com/AlexeyAB/darknet/issues/4147#issuecomment-548481948

**Comment by WongKinYiu** [comment]
i ll get free gpus after 1/24 i think.

**Comment by glenn-jocher** [comment]
@WongKinYiu @AlexeyAB hi guys. I'm going to try to implement matrix nets in our https://github.com/ultralytics/yolov3 repo. 

To be able to use these updated cfg files it looks like I need to parse `stride_x` and `stride_y` info and then implement these special convolutions. Just thinking out loud here I need to see if PyTorch has this capability, as right now we pass a single stride scalar to the Conv2d module. Maybe I can do this by passing an array of strides instead.

```bash
        if mdef['type'] == 'convolutional':
            bn = int(mdef['batch_normalize'])
            filters = int(mdef['filters'])
            kernel_size = int(mdef['size'])
            pad = (kernel_size - 1) // 2 if int(mdef['pad']) else 0
            modules.add_module('Conv2d', nn.Conv2d(in_channels=output_filters[-1],
                                                   out_channels=filters,
                                                   kernel_size=kernel_size,
                                                   stride=int(mdef['stride']),
                                                   padding=pad,
                                                   bias=not bn))
```

**Comment by WongKinYiu** [comment]
@glenn-jocher Hello,

The stride in Pytorch can be tuple.

**Comment by glenn-jocher** [comment]
@WongKinYiu right, it looks like I should update `stride=int(mdef['stride'])` to this. I think this is the proper order right? Do the `upsample` layers need to be modified also or is the only change required in the `Conv2d()` layers??

`stride=(int(mdef['stride_y']),  int(mdef['stride_x']))`

https://pytorch.org/docs/stable/nn.html#conv2d
<img width="590" alt="Screenshot 2019-11-12 at 17 17 16" src="https://user-images.githubusercontent.com/26833433/68724439-55ab7f00-0570-11ea-8bd5-b4b1d010690d.png">

**Comment by WongKinYiu** [comment]
i think modified conv2d is enough for matrixnet.

**Comment by glenn-jocher** [comment]
@WongKinYiu ok, I've updated our https://github.com/ultralytics/yolov3 repo to handle cfg [convolutional] layers with either `stride` or `stride_x` and `stride_y`. What is the best matrixnets cfg file you would recommend to try as a yolov3-spp.cfg replacement on COCO?

Our current yolov3-spp.cfg results on ultralytics/yolov3 are about 0.40mAP@0.5:0.95 and 0.60mAP@0.5 (see https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-552271610), so this is the benchmark I want to test against and improve.

**Comment by WongKinYiu** [comment]
@glenn-jocher Hello,

I haven't design matrixnets for big models.
I thank I need 2~3 days to design it.

By the way, I have cfg file for another model, but it can not be released currently.
I think it can easily get better result than https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-552271610 using your training skill. (but it need a little bit more inference time than YOLOv3-SPP on Darknet)
If your are interested in it, please tell me then I can send it to you.

**Comment by isgursoy** [comment]
@WongKinYiu i would have a look

**Comment by WongKinYiu** [comment]
@glenn-jocher Hello,

https://github.com/ruinmessi/ASFF

They use
* Region proposal by guided anchoring (CVPR 2019)
*  Unitbox: An advanced object detection network (ACMMM 2016)
*  Bag of freebies for training object detection neural networks (arXiv 2019)

Then AP is improve 5.8%
![image](https://user-images.githubusercontent.com/12152972/69711562-3f8aea80-113c-11ea-86b6-0c4c608fa7fa.png)

With their proposed ASFF
AP is improve 7.7%
![image](https://user-images.githubusercontent.com/12152972/69711701-752fd380-113c-11ea-803a-8823328823a2.png)

After apply
* Dropblock: A regularization method for convolutional networks (NeurIPS 2018)
* Receptive field block net for accurate and fast object detection (ECCV 2018)

AP is improve 9.4%
![image](https://user-images.githubusercontent.com/12152972/69711906-d3f54d00-113c-11ea-8425-9edd9b56f90e.png)

Their code is written in PyTorch, maybe you can add these training trick to your repo.

**Comment by AlexeyAB** [comment]
@WongKinYiu Yes, there is already issue for ASFF: https://github.com/AlexeyAB/darknet/issues/4382

1. As I understand the greatest increase in accuracy was given by **MixUp** (BoF) `+4.2AP@0.5...0.95` and `+5.6%AP@70` I implemented it for Detector, but I think I should implement it for Classifier too (since MixUp gives the greatest increase in accuracy in classification)  https://github.com/AlexeyAB/darknet/issues/3272

2. Also there is used **synchronized batch normalization** technique for increasing mini_batch size (in the same was is in MegDet: A Large Mini-Batch Object Detector - paper https://arxiv.org/abs/1711.07240v4 ) - I implemented GPU-processing on CPU-RAM, so you can use a lot of cheap CPU RAM for increasing **mini_batch** up to 32-128 on Yolov3, and 128-512 on Classification networks: https://github.com/AlexeyAB/darknet/issues/4386

3. **Dropblock** + **Receptive field block** gives `+1.7% AP@0.5...0.95`

So **ASFF** gives only  `+1.8% AP@0.5...0.95` and `1.5% AP@0.5` and `2.5% AP@.07`

![image](https://user-images.githubusercontent.com/4096485/69716408-fe361300-111a-11ea-9cc1-b6aa62154a9c.png)

**Comment by glenn-jocher** [comment]
@WongKinYiu @AlexeyAB thanks, I'll have to read the paper. I wonder why they used yolov3 and not yolov3-spp as the baseline. @WongKinYiu did you make any progress on matrix net cfgs for yolov3-spp? 

My latest yolov3-spp.cfg trainings are at 0.404mAP@0.5:0.95 and 0.601mAP@0.5 at 608 using multiscale 288-640. I'm running more models currently that should be done in about a week (new hyps, more multiscale, swish, etc.)

**Comment by glenn-jocher** [comment]
@AlexeyAB oh BTW, what's the easiest way to train on COCO2017? Is there a darknet ready version out there somewhere?

**Comment by AlexeyAB** [comment]
@glenn-jocher 

You mean, is there a script that converts MSCOCO 2017 annotations into yolo format?

Did you try this project? https://bitbucket.org/yymoto/coco-to-yolo/src/master/
https://commecica.com/2018/07/18/starting-with-darknet-yolo-coco/

A pre-built JAR file can be found here: http://commecica.com/wp-content/uploads/2018/07/cocotoyolo.jar

Run

`java -jar cocotoyolo.jar "coco/annotations/instances_train2017.json" "/usr/home/madmax/coco/images/train2017/" "all" "coco/yolo"`

and
`java -jar cocotoyolo.jar "coco/annotations/instances_val2017.json" "/usr/home/madmax/coco/images/val2017/" "all" "coco/yolo"`

**Comment by WongKinYiu** [comment]
@glenn-jocher 

i dont know why they use yolov3 as baseline.
but their model has spp layer.
https://github.com/ruinmessi/ASFF/blob/master/models/yolov3_asff.py#L35
![image](https://user-images.githubusercontent.com/12152972/69768285-f75edd00-11ba-11ea-93a3-68a56ed76ae3.png)

**Comment by glenn-jocher** [comment]
@WongKinYiu oh thats kind of confusing. @AlexeyAB ok thanks for the code, I'll try to convert COCO2017 to yolo format so I can get mAPs on it also. 

I wonder how COCO2014 mAPs relate to COCO2017 mAPs.

**Comment by WongKinYiu** [comment]
the test-dev set of COCO2014 and COCO2017 are same.
the main difference is training data.
" Finally, new in 2017 we are releasing 120K unlabeled images from COCO that follow the same class distribution as the labeled images; this may be useful for semi-supervised learning on COCO."
http://cocodataset.org/#download

**Comment by glenn-jocher** [comment]
@WongKinYiu yes I think you are right. The darknet training set (2014) is 117k images, and tests on 5k images. The 2017 set I believe trains on 120k images and tests on 20k. So testing will take 4X longer on 2017, but I believe most of the training data is the same.

**Comment by glenn-jocher** [comment]
I've updated my mAP section, so the jump between YOLOv3 and YOLOv3-SPP is more clear. We are not at the level of ASFF yet, but it's encouraging to see that we are not too far away either. 

https://github.com/ultralytics/yolov3#map

<i></i>                      |img-size |COCO mAP<br>@0.5...0.95 |COCO mAP<br>@0.5 
---                          | ---         | ---         | ---
YOLOv3-tiny<br>YOLOv3<br>YOLOv3-SPP<br>**YOLOv3-SPP ultralytics** |320 |14.0<br>28.7<br>30.5<br>**35.2** |29.0<br>51.5<br>52.3<br>**53.9**
YOLOv3-tiny<br>YOLOv3<br>YOLOv3-SPP<br>**YOLOv3-SPP ultralytics** |416 |16.0<br>31.1<br>33.9<br>**38.8** |32.9<br>55.3<br>56.8<br>**58.7**
YOLOv3-tiny<br>YOLOv3<br>YOLOv3-SPP<br>**YOLOv3-SPP ultralytics** |608 |16.6<br>33.0<br>37.0<br>**40.4** |35.5<br>57.9<br>**60.6**<br>60.1

The main takeaway for me is their Figure 1. This shows that while some of the new work that's coming out claims very high mAPs, they do it at the expense of inference time, so in this sense I like the AFSS approach of adding to YOLOv3 to obtain results with minimal hits to FPS.
<img width="325" alt="Screen Shot 2019-11-27 at 4 03 08 PM" src="https://user-images.githubusercontent.com/26833433/69771061-7dcddd00-112f-11ea-85a5-a9d862dfc6c1.png">

**Comment by WongKinYiu** [comment]
@glenn-jocher yes,

there are too many tricks which are hard to implemented in darknet...
maybe i ll test new methods on pytorch then feedback it to @AlexeyAB in the future.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

> there are too many tricks which are hard to implemented in darknet...

What features do you mean?


I added several fixes, so ASFF can be implemented there - I hope there are no bugs: https://github.com/AlexeyAB/darknet/commit/68db8efb27170b411481e8765558d89b45635101

ASSF - like:
```
[route]
layers=22,33,44 # 3-layers which are already resized to the same WxHxC

[convolutional]
stride=1
size=1
filters=3
activation=normalize_channels # may be before this activation we should do ReLU or leaky-ReLU???

[route]
layers=-1
group_id=0
groups=3

[scale_channels]
from=22
scale_wh=1

[route]
layers=-3
group_id=1
groups=3

[scale_channels]
from=33
scale_wh=1


[route]
layers=-5
group_id=2
groups=3

[scale_channels]
from=44
scale_wh=1

[shortcut]
from=-3
activation=linear

[shortcut]
from=-6
activation=linear


```

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

For example, Autoaugment, [fast-autoaugment](https://github.com/kakaobrain/fast-autoaugment), autograd, deformable conv, and **multi-task training**...

In my project, I need do detection, segmentation, and tracking simultaneously.
[link1](https://github.com/Zhongdao/Towards-Realtime-MOT), [link2](https://www.youtube.com/watch?v=nndFtIPMy20&feature=youtu.be), [link3](https://github.com/dbolya/yolact)
but modify the data loader and build a multi-task detector-segmentor-tracker are difficult to me.

by the way, I will create a ASFF-like cfg in thease days.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

Which of these tasks do you consider the most promising?

Yes, instance segmentation - YOLACT still in todo: https://github.com/AlexeyAB/darknet/projects/6 Like as tracker: https://github.com/AlexeyAB/darknet/issues/3042

I will think can I do fast "Joint Detection and Embedding for fast multi-object tracking"

Just now the simplest way just to add Monving to data augmentation, so we will can train conv-LSTM models on non-sequential datasets like MS COCO, so it will increase accuracy of detection on video, but will still require additional tracker for track_id's.

The highest priorities were: https://github.com/AlexeyAB/darknet/issues/4264  https://github.com/AlexeyAB/darknet/issues/4346  https://github.com/AlexeyAB/darknet/issues/4382  https://github.com/AlexeyAB/darknet/issues/4203 https://github.com/AlexeyAB/darknet/issues/3830

Do you think Deformable Conv better than Deformable Kernels ?  https://github.com/AlexeyAB/darknet/issues/4066

**Comment by WongKinYiu** [comment]
@AlexeyAB 

For me, "Joint Detection and Embedding for fast multi-object tracking".
For improving YOLOv3, ASFF and tricks used in the ASFF paper.

Oh, I treat deformable kernel as deformable conv v3.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

Did you try to compare `CSPNet` with `ResNet101-deformable-conv-TridentNet` (FPS / accuracy)? https://arxiv.org/abs/1901.01892v1 issue https://github.com/AlexeyAB/darknet/issues/3363

How slow are deformable convolutions?

It addes only +~1% AP but may be it is very slow?

![image](https://user-images.githubusercontent.com/4096485/69867269-5dc92580-12b7-11ea-9baf-35f7c62b824d.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB 

I have not try it.
If darknet support deformable conv, i can train the model.

In ICCV, the author of tridentnet show the fast version of tridentnet.
They train on multiple branch, and test on single branch.
https://github.com/facebookresearch/detectron2/tree/master/projects/TridentNet

**Comment by AlexeyAB** [comment]
@WongKinYiu 

They don't use Deformable-conv in the `Resnet101-TridentNet-Fast scale-aware`. 

`Resnet152-TridentNet` is already implemented in the Darknet. https://github.com/AlexeyAB/darknet/blob/master/cfg/resnet152_trident.cfg 

But we should implement `Resnet101-TridentNet` rather than Resnet152: https://github.com/AlexeyAB/darknet/issues/3363

To implement `Resnet101-TridentNet-Fast scale-aware` we should:
* Use ResNet101 as backbone: https://github.com/AlexeyAB/darknet/blob/master/cfg/resnet101.cfg and weights https://pjreddie.com/media/files/resnet101.weights
* Implement 3-branches with shared weights `share_index=` and different `dilation=` as in https://github.com/AlexeyAB/darknet/blob/master/cfg/resnet152_trident.cfg
* remove `masks=` in each [yolo] layer, for **enlarging the scale-aware range of the major branch to incorporate objects of all scales**
* train this model on mscoco by using pre-trained weights file https://pjreddie.com/media/files/resnet101.weights
* then just remove 1st and 3rd branches, keep branch-2

It will be `Resnet101-TridentNet-Fast scale-aware` 


----

They use `ResNet-101` (40.6% AP) rather than `ResNet-101-Deformable` (41.8% AP), because Deformable-conv is very slow for Fast-network.

 **`ResNet-101` (40.6% AP)** vs **`ResNet-101-Deformable` (41.8% AP)**
![image](https://user-images.githubusercontent.com/4096485/69891266-4c5e3880-130c-11ea-971c-e1d2461b16d3.png)

----

 **Resnet101-TridentNet-Fast scale-aware**
![image](https://user-images.githubusercontent.com/4096485/69891271-5849fa80-130c-11ea-8d60-7da830fa23d5.png)


----

**Comment by Kyuuki93** [comment]
> ASSF - like:
> 
> ```
> [route]
> layers=22,33,44 # 3-layers which are already resized to the same WxHxC
> 
> [convolutional]
> stride=1
> size=1
> filters=3
> activation=normalize_channels # ReLU is integrated to activation=normalize_channels
> 
> [route]
> layers=-1
> group_id=0
> groups=3
> 
> [scale_channels]
> from=22
> scale_wh=1
> 
> [route]
> layers=-3
> group_id=1
> groups=3
> 
> [scale_channels]
> from=33
> scale_wh=1
> 
> 
> [route]
> layers=-5
> group_id=2
> groups=3
> 
> [scale_channels]
> from=44
> scale_wh=1
> 
> [shortcut]
> from=-3
> activation=linear
> 
> [shortcut]
> from=-6
> activation=linear
> ```

@AlexeyAB
As I understand it ,  

```
[route]
layers = 22, 33, 44
```

 were resized from the those 3 layers,

<img width="457" alt="image" src="https://user-images.githubusercontent.com/26849480/69949337-43e44880-152c-11ea-9bb6-e524fe1f19e6.png">

 that's right?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

They released their [long paper](https://arxiv.org/abs/2001.03194) and [code](https://github.com/arashwan/matrixnet).

![image](https://user-images.githubusercontent.com/12152972/72433213-cfabee80-37d3-11ea-80b0-ac11e4a9cb4e.png)

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks!

MatrixNet achieves only 3-4 FPS, while CSPResNeXt-50-PANet-SPP-optimal achieves 30-40 FPS with ~ the same AP.

I added something like CenterNet that uses 3 x [Gaussian_yolo] layers (left-top, right-bottom, center) for each final resolution. https://github.com/AlexeyAB/darknet/issues/3229#issuecomment-569412122
You can try to use it with MatrixNet structure.

**Comment by AlexeyAB** [comment]
@WongKinYiu Also I fixed a bug with Tensor Cores, and it seems any groups value larger than `groups=1` slows down inference speed, so `CSPDarknet53` and `CSPDarkNet53-PANet-SPP` should be the best model even for 608x608 if **Tensor Cores** are used.

**Comment by WongKinYiu** [comment]
@AlexeyAB Thanks a lot,

I am doing ablation study of `CSPResNeXt50-PANet-SPP-Optimal`, so there is no available large-ram-gpus recently.
I will test new methods using small models on small dataset.

Yes, `CSPDarknet53-PANet-SPP` gets best AP and FPS currently, I am developing new methods based on `CSPDarknet53`. 

I find that the bad detection usually appears on overlapped objects in my project.
If your focused task also has this kind of problem, do you think the `Soft-IoU` layer [(cvpr19-code)](https://github.com/eg4000/SKU110K_CVPR19) can be combined with `YOLOv3` since you already introduced the concept of `CenterNet` into `YOLOv3`?

It seems `ASFF` and `Bi-FPN` are very easy to get NaN even combine with original `Darknet53`.
I am trying to solve the problems.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

* Do you mean that you trained the ASFF-model which @Kyuuki93 provided and got Nan? https://github.com/AlexeyAB/darknet/issues/4382

* Did you train `yolov3-spp-asff-it-rfb.cfg.txt` on MS COCO with ASFF+RFB+DropBlock?

* Do you get a slowdown in training with the current dropblock implementation?

* Did you train this BiFPN model `darknet53-bifpn3.cfg.txt` on MS COCO and get Nan? https://github.com/AlexeyAB/darknet/issues/4662#issuecomment-573274750



> I find that the bad detection usually appears on overlapped objects in my project.
> If your focused task also has this kind of problem, do you think the Soft-IoU layer (cvpr19-code) can be combined with YOLOv3 since you already introduced the concept of CenterNet into YOLOv3?

I will think about it. I created an issue: https://github.com/AlexeyAB/darknet/issues/4701

Also what do you think about Repulsion Loss: Detecting Pedestrians in a Crowd (CVPR 2018) - without NMS at all ? https://arxiv.org/abs/1711.07752v2 and https://github.com/AlexeyAB/darknet/issues/3113

**Comment by WongKinYiu** [comment]
@AlexeyAB 

* yes, it gets nan after 140k epochs when learning rate is 0.0005.

* i do not train with dropblock, but all of asff, rfb, asff+rfb are easy to get nan when use ciou or giou loss.
mse loss seems more stable for asff and rfb.

* yes, training with dropblock still slower than without dropblock.

* yes, it gets nan when all of other hyper-parameters are set as same as csresnext50-panet-spp-optimal.
currently i reduce the learning rate to 0.001 and re-train it.

* thanks

* yes, if there is no nms is the best solution since nms does not considerate any semantic information.

**Comment by Kyuuki93** [comment]
@WongKinYiu What normalizer you used in ASFF module? In my experiments, ASFF got NaN when use relu for weights normalize and change to softmax will resolve this, the reason could be something like dead neurons which happens in relu

**Comment by AlexeyAB** [comment]
@Kyuuki93 
@WongKinYiu Yes, you should use `activation=normalize_channels_softmax` instead of `activation=normalize_channels` in conv-layer before `[scale_channels] scale_wh=1` in ASFF:
```
[convolutional]
stride=1
size=1
filters=3
activation=normalize_channels_softmax
#activation=normalize_channels            # don't use
```


> yes, training with dropblock still slower than without dropblock.

Is it about ~2x slower?

**Comment by AlexeyAB** [comment]
@WongKinYiu Also use `batch_normalization=1` in RFB-block

**Comment by AlexeyAB** [comment]
@WongKinYiu 

Try to use `max_delta=10` in `[yolo]` layers to avoid `Nan`.

* Did you solve issue with Nan for ASFF and RFB ?

* I added `max_delta` param for `[yolo]` and `[Gaussian_yolo]`, so you can use `max_delta=10` (preferably) or `max_delta=100` it should limit `delta` to avoid Nan

* Also you can try to use `scale_x_y=1.2` or higher, it may allow you to use lower max_delta, but it can lead to lower AP (I don't know)

**Comment by WongKinYiu** [comment]
@AlexeyAB Thanks a lot

nan is not yet solved for asff model (rfb without asff do not get nan)
i m in an one week vocation, will start training at 1/30.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

I added `[convolutional] activation=normalize_channels_softmax_maxval` https://github.com/AlexeyAB/darknet/commit/298805cefebd0e71274bc12c068384703c9684b4

Try to train with
`[convolutional] activation=normalize_channels_softmax_maxval`
instead of 
`[convolutional] activation=normalize_channels_softmax`
for ASFF  in  3 [conv] layers

**Comment by WongKinYiu** [comment]
@AlexeyAB OK, thanks a lot.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

I just checked your cfg-files:

* Download the latest Darknet

* set `batch_normalize=1` instead of `batch_normalize=0` in RFB-block

* add `max_delta=10` for each `[yolo]` layer (the closer to 0  value - the more stable the training)

* Try both, either 
`activation=normalize_channels_softmax_maxval` or
`activation=normalize_channels_softmax` (I don't know which is more stable)

**Comment by WongKinYiu** [comment]
* Download the latest Darknet

* set `batch_normalize=1` instead of `batch_normalize=0` in RFB-block

* add `max_delta=10` for each `[yolo]` layer (the closer to 0 value - the more stable the training)

* `activation=normalize_channels_softmax`

still not stable enough for `asff` and `asff+rfb`, now reduce learning rate and retraining.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

Update your code at least to this fix: https://github.com/AlexeyAB/darknet/commit/2862b28dc7327b2ec7085da0f06b665e48c71b50

> still not stable enough for asff and asff+rfb,

At what number of iterations it goes to Nan?


Also try 
`activation=normalize_channels_softmax_maxval`

**Comment by WongKinYiu** [comment]
@AlexeyAB 

I use https://github.com/AlexeyAB/darknet/tree/61499b27a4e24656a0f84bb83b92df95b0917f74 for training.

I already clean the results, but it got nan at < 35k epochs with `lr=0.001`.
